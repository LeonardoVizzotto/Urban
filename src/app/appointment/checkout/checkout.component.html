<app-container *ngIf="selectedSlot && selectedWorker">
  <app-spacer size="large"></app-spacer>
  <img
    [src]="selectedWorker.profilePicture?.large"
    alt="Worker profile picture"
  />
  <app-spacer size="small"></app-spacer>
  <app-rating [rating]="selectedWorker.rating"></app-rating>
  <app-spacer size="medium"></app-spacer>
  <span
    >You selected <b>{{ selectedWorker.name }}</b> to be your therapist at
    <b>{{ selectedSlot.localisedTime }}</b> today</span
  >
  <app-spacer size="small"></app-spacer>
  <span
    ><b> Price: {{ selectedSlot.price }}</b>
  </span>
  <app-spacer size="large"></app-spacer>
  <div class="button-group">
    <app-button [appearance]="'secondary-ghost'" (onClick)="changeTherapist()"
      >Change choices</app-button
    >
    <app-button [appearance]="'primary'" (onClick)="bookAppointment()"
      >Book now</app-button
    >
  </div>
</app-container>
