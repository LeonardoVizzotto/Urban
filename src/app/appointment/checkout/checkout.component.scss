img {
  border-radius: 50%;
}

span {
    text-align: center;
}

.button-group {
  width: 60%;
  display: flex;
  gap: 20px;
  > * {
    flex: 1 1 0;
  }
}
