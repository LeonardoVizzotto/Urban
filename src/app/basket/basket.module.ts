import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BasketComponent } from './basket/basket.component';
import { RouterModule, Routes } from '@angular/router';
import { SharedModule } from '../shared/shared.module';
import { AppointmentItemComponent } from './appointment-item/appointment-item.component';

const routes: Routes = [
  {
    path: '',
    component: BasketComponent,
  },
];

@NgModule({
  declarations: [BasketComponent, AppointmentItemComponent],
  imports: [CommonModule, RouterModule.forChild(routes), SharedModule],
})
export class BasketModule {}
