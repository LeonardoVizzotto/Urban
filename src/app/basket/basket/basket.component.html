<app-container *ngIf="appointments$ | async; let appointments">
  <div *ngIf="appointments.length; else empty" class="appointment-container">
    <app-appointment-item
      *ngFor="let appointment of appointments; let i = index"
      [appointment]="appointment"
      (removeAppointment)="removeAppointment(i)"
    >
    </app-appointment-item>
  </div>
  <ng-template #empty>
    <span>Looks like your basket is empty</span>
  </ng-template>

  <app-spacer size="large"></app-spacer>
  <app-button (onClick)="startBooking()" [appearance]="'primary-ghost'"
    >Start Booking</app-button
  >
</app-container>
