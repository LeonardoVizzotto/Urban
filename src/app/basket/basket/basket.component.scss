.appointment-container {
    width: 100%;
}