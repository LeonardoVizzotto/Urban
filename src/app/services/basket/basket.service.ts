import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { Slot } from '../slots/model';
import { Worker } from '../workers/model';
import { Appointment } from './model';

@Injectable({
  providedIn: 'root',
})
export class BasketService {
  private _appointments: Appointment[] = [];
  private _selectedSlot?: Slot;
  private _selectedWorker?: Worker;

  private appointmentsSubject$ = new BehaviorSubject<Appointment[]>(
    this._appointments
  );
  public appointments$ = this.appointmentsSubject$.asObservable();

  constructor() {
  }

  public selectSlot(slot: Slot) {
    this._selectedSlot = slot;
  }

  public get selectedSlot() {
    return this._selectedSlot;
  }

  public selectWorker(worker: Worker) {
    this._selectedWorker = worker;
  }

  public reset() {
    this._selectedSlot = void 0;
    this._selectedWorker = void 0;
  }

  public get selectedWorker() {
    return this._selectedWorker;
  }

  public bookAppointment() {
    if (!this.selectedSlot || !this.selectedWorker) {
      return;
    }

    this._appointments.push({
      slot: this.selectedSlot,
      worker: this.selectedWorker,
    });

    this.appointmentsSubject$.next([...this._appointments]);
    this.reset();
  }

  public removeAppointment(index: number) {
    this._appointments.splice(index, 1);
    this.appointmentsSubject$.next([...this._appointments]);
  }
}
