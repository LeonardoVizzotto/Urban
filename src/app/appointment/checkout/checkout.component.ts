import { Component, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { ROUTES } from 'src/app/routes.constants';
import { BasketService } from 'src/app/services/basket/basket.service';
import { Slot } from 'src/app/services/slots/model';
import { ToastService } from 'src/app/services/toast/toast.service';
import { Worker } from 'src/app/services/workers/model';

@Component({
  selector: 'app-checkout',
  templateUrl: './checkout.component.html',
  styleUrls: ['./checkout.component.scss'],
})
export class CheckoutComponent implements OnInit {
  constructor(
    private basketService: BasketService,
    private router: Router,
    private toastService: ToastService
  ) {}

  public selectedSlot?: Slot;
  public selectedWorker?: Worker;

  ngOnInit(): void {
    this.selectedSlot = this.basketService.selectedSlot;
    this.selectedWorker = this.basketService.selectedWorker;

    if (!this.selectedSlot || !this.selectedWorker) {
      this.router.navigate([ROUTES.SLOTS]);
    }
  }

  public changeTherapist() {
    this.basketService.reset();
    this.router.navigate([ROUTES.SLOTS]);
  }

  public bookAppointment() {
    this.basketService.bookAppointment();
    this.toastService.success(
      `Successfully booked an appointment with ${this.selectedWorker?.name} for today ${this.selectedSlot?.localisedTime}`
    );
    this.router.navigate([ROUTES.BASKET]);
  }
}
