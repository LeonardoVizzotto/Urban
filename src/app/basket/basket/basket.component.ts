import { Component, OnInit } from '@angular/core';
import { Router } from '@angular/router';
import { ROUTES } from 'src/app/routes.constants';
import { BasketService } from 'src/app/services/basket/basket.service';

@Component({
  selector: 'app-basket',
  templateUrl: './basket.component.html',
  styleUrls: ['./basket.component.scss'],
})
export class BasketComponent {
  public appointments$ = this.basketService.appointments$;
  constructor(private basketService: BasketService, private router: Router) {}

  public removeAppointment(index: number) {
    this.basketService.removeAppointment(index);
  }

  public startBooking() {
    this.router.navigate([ROUTES.SLOTS]);
  }
}
