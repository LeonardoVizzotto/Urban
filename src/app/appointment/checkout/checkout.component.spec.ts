import { ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { BasketService } from 'src/app/services/basket/basket.service';
import { SharedModule } from 'src/app/shared/shared.module';
import { mockProviders } from 'src/app/test/utils';

import { CheckoutComponent } from './checkout.component';

describe('CheckoutComponent', () => {
  let component: CheckoutComponent;
  let fixture: ComponentFixture<CheckoutComponent>;
  const { providers, router } = mockProviders();

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [SharedModule, RouterTestingModule],
      declarations: [CheckoutComponent],
      providers,
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CheckoutComponent);
    component = fixture.componentInstance;
    router.navigate.mockReset();
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  it('should navigate to SLOTS if no slot has been selected', () => {
    const basketService = TestBed.inject(BasketService);
    basketService.reset();
    fixture.detectChanges();
    expect(router.navigate).toHaveBeenCalled();
  });

  it('should match snapshot if a slot and a worker have been selected', () => {
    const basketService = TestBed.inject(BasketService);
    basketService.selectSlot({
      id: 1,
      localisedTime: '12:00',
      price: 'Â£81.00',
    });
    basketService.selectWorker({
      id: 1,
      name: 'John Doe',
      rating: '2.9',
      isNew: false,
      profilePicture: {
        large: 'https://randomuser.me/api/portraits/men/22.jpg',
        medium: 'https://randomuser.me/api/portraits/med/men/22.jpg',
        thumbnail: 'https://randomuser.me/api/portraits/thumb/men/22.jpg',
      },
    });
    fixture.detectChanges();
    expect(router.navigate).not.toHaveBeenCalled();
    expect(fixture).toMatchSnapshot();
  });
});
