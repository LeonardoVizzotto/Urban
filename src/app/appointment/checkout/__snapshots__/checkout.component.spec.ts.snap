// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutComponent should match snapshot if a slot and a worker have been selected 1`] = `
<app-checkout
  basketService={[Function BasketService]}
  router={[Function Object]}
  selectedSlot={[Function Object]}
  selectedWorker={[Function Object]}
  toastService={[Function ToastService]}
>
  <app-container>
    <app-spacer
      aria-hidden="true"
      class="[object Object]"
      size="large"
    />
    <img
      alt="Worker profile picture"
      src="https://randomuser.me/api/portraits/men/22.jpg"
    />
    <app-spacer
      aria-hidden="true"
      class="[object Object]"
      size="small"
    />
    <app-rating>
      <div
        aria-label="Rating of this worker is 2.9 out of 5."
        class="rating"
        style="--rating: 2.9;"
      />
    </app-rating>
    <app-spacer
      aria-hidden="true"
      class="[object Object]"
      size="medium"
    />
    <span>
      You selected 
      <b>
        John Doe
      </b>
       to be your therapist at 
      <b>
        12:00
      </b>
       today
    </span>
    <app-spacer
      aria-hidden="true"
      class="[object Object]"
      size="small"
    />
    <span>
      <b>
         Price: Â£81.00
      </b>
    </span>
    <app-spacer
      aria-hidden="true"
      class="[object Object]"
      size="large"
    />
    <div
      class="button-group"
    >
      <app-button>
        <button
          class="button -secondary-ghost"
          type="button"
        >
          Change choices
        </button>
      </app-button>
      <app-button>
        <button
          class="button -primary"
          type="button"
        >
          Book now
        </button>
      </app-button>
    </div>
  </app-container>
</app-checkout>
`;
